pennylane_lightning_kokkos/src/simulator/StateVectorKokkos.hpp

```diff
@@ -445,6 +446,99 @@ template <class Precision> class StateVectorKokkos {
         }
     };
 
+    /**
+     * @brief Utility method for samples.
+     *
+     * @param num_samples Number of Samples
+     *
+     * @return Kokkos::View<size_t *> to the samples.
+     * Each sample has a length equal to the number of qubits. Each sample can
+     * be accessed using the stride sample_id*num_qubits, where sample_id is a
+     * number between 0 and num_samples-1.
+     */
+
+    auto generate_samples(size_t num_samples)->Kokkos::View<size_t *> {
+
+        const size_t num_qubits = getNumQubits(); // no need for dec
+
+        const size_t N = getLength();
+
+        const size_t Nsqrt =
+            static_cast<size_t>(std::ceil(std::sqrt(static_cast<double>(N))));
+
+        Kokkos::View<Kokkos::complex<Precision> *> arr_data = getData();
+
+        Kokkos::View<Precision *> probabilities("probabilities", N);
+
+        Kokkos::View<Precision *> cld("cld", N);
+
+        Kokkos::View<size_t *> samples("num_samples", num_samples * num_qubits);
+
+        Kokkos::parallel_for(
+            Kokkos::RangePolicy<KokkosExecSpace>(0, N),
+            KOKKOS_LAMBDA(const size_t &i) {
+                probabilities[i] = imag(arr_data[i]) * imag(arr_data[i]) +
+                                   real(arr_data[i]) * real(arr_data[i]);
+            });
+
+        Kokkos::parallel_for(
+            Kokkos::RangePolicy<KokkosExecSpace>(0, Nsqrt),
```

Comment:
The computation of cld can be combined with the computation of probabilities I think, this would make it perhaps faster. In fact, the probabilities array can be completely removed I think

Reply:
> The computation of cld can be combined with the computation of probabilities I think, this would make it perhaps faster. In fact, the probabilities array can be completely removed I think

Agree!

Reply:
btw this all can probably be done with Kokkos parallel_scan function

Reply:
done!